src/transformers/models/moshi/modeling_moshi.py

```diff
@@ -0,0 +1,2253 @@
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        **kwargs,
+    ) -> torch.LongTensor:
+        """
+        # TODO: modify
+        Generates sequences of token ids for models with a language modeling head.
+
+        Parameters:
+            inputs (`torch.Tensor` of varying shape depending on the modality, *optional*):
+                The sequence used as a prompt for the generation or as model inputs to the encoder.
+            kwargs (`Dict[str, Any]`, *optional*):
+                Remaining dictionary of keyword arguments that are passed to the `generate` method. Refers to the 
+                original [`generate` docstrings](https://huggingface.co/docs/transformers/main/en/main_classes/text_generation#transformers.GenerationMixin.generate)
+                for more information on how to use them.
+                Note that keywords with a *depth_* prefix will be input for the `generate` method of the
+                depth decoder. Otherwise, the latter will use its default generation config.
+    
+        Return: # TODO
+            [`~utils.ModelOutput`] or `torch.LongTensor`: A [`~utils.ModelOutput`] (if `return_dict_in_generate=True`
+            or when `config.return_dict_in_generate=True`) or a `torch.FloatTensor`.
+
+                If the model is *not* an encoder-decoder model (`model.config.is_encoder_decoder=False`), the possible
+                [`~utils.ModelOutput`] types are:
+
+                    - [`~generation.GenerateDecoderOnlyOutput`],
+                    - [`~generation.GenerateBeamDecoderOnlyOutput`]
+
+                If the model is an encoder-decoder model (`model.config.is_encoder_decoder=True`), the possible
+                [`~utils.ModelOutput`] types are:
+
+                    - [`~generation.GenerateEncoderDecoderOutput`],
+                    - [`~generation.GenerateBeamEncoderDecoderOutput`]
+        """
+        inputs_embeds, moshi_audio_codes = self._prepare_inputs_embeds_for_generation(
+            input_ids=input_ids,
+            user_input_values=user_input_values,
+            user_audio_codes=user_audio_codes,
+            moshi_input_values=moshi_input_values,
+            moshi_audio_codes=moshi_audio_codes,
+            inputs_embeds=inputs_embeds,
+        )
+        
+        self.generated_audio_codes = moshi_audio_codes
+        
+        outputs = super().generate(inputs_embeds=inputs_embeds, **kwargs)
+
+        # check if outputs is a dict or a Tensor (depending on unaccessed `generation_config.return_dict_in_generate`)
+        if isinstance(outputs, torch.Tensor):
+            output_text_ids = outputs
+        else:
+            output_text_ids = outputs.sequences
+            
+        output_audio_codes = self.generated_audio_codes
+
+        
+        output_values = self.audio_encoder.decode(
+            output_audio_codes,
+        ).audio_values
+      
+        
+        return output_text_ids, output_values
```

Comment:
I actually would like to allow dynamic outputs depending on the type of generation (beam, sample) etc., do you think I can do a nested `ModelOutput` ?

Reply:
My suggestion would be to make it as close as possible to the return structure from the original `generate`. Users transitioning from other models to `moshi` would then have as little friction as possible 🤗